desc/optimize/optimizer.py

```diff
@@ -146,15 +147,45 @@ def optimize(  # noqa: C901 - FIXME: simplify this
             linear_constraints = maybe_add_self_consistency(linear_constraints)
         if len(linear_constraints):
             objective = LinearConstraintProjection(objective, linear_constraints)
+            if nonlinear_constraint is not None:
+                nonlinear_constraint = LinearConstraintProjection(
+                    nonlinear_constraint, linear_constraints
+                )
         if not objective.built:
             objective.build(eq, verbose=verbose)
+        if nonlinear_constraint is not None and not nonlinear_constraint.built:
```

Comment:
if the nonlinear constraint is not None at this point, doesn't that mean it wasn't wrapped properly? I am a bit confused at the logic

Reply:
My confusion is that in line 141
```python
 objective, nonlinear_constraint = _maybe_wrap_nonlinear_constraints(
            objective, nonlinear_constraint, self.method, options
        )
```
the `_maybe_wrap_nonlinear_constraints`  if there is  non-None `nonlinear_constraint` passed in, seems to return objective, None

```python
if wrapper is not None and wrapper.lower() in ["prox", "proximal"]:
        perturb_options = options.pop("perturb_options", {})
        solve_options = options.pop("solve_options", {})
        objective = ProximalProjection(
            objective,
            constraint=nonlinear_constraint,
            perturb_options=perturb_options,
            solve_options=solve_options,
        )
        nonlinear_constraint = None
    return objective, nonlinear_constraint
```
But ok maybe I see my confusion. This is just checking if the wrapper is proximal projection, at which point we fold the nonlinear constraint into the ProximalProjection objective which takes care of it after each solver iteration, so we no longer need to consider the nonlinear constraint. That is only something optimizer needs to consider if we are using like a constrained optimizer with a nonlinear constraint.

Reply:
Yes exactly